File: backend/app/core/response.py
from typing import Any

from fastapi.responses import JSONResponse


def ok(data: Any = None, meta: dict[str, Any] | None = None) -> dict[str, Any]:
    payload: dict[str, Any] = {"code": 0, "message": "success", "data": data}
    if meta is not None:
        payload["meta"] = meta
    return payload


def fail(code: int, message: str) -> dict[str, Any]:
    return {"code": code, "message": message, "data": None}


class ApiError(Exception):
    def __init__(self, code: int, message: str, status_code: int = 400):
        self.code = code
        self.message = message
        self.status_code = status_code
        super().__init__(message)


def error_response(error: ApiError) -> JSONResponse:
    return JSONResponse(status_code=error.status_code, content=fail(error.code, error.message))

File: backend/internal/models/response.go
package models

import (
	"encoding/json"
	"net/http"
)

type PageMeta struct {
	Total    int64 `json:"total"`
	Page     int   `json:"page"`
	PageSize int   `json:"page_size"`
	HasNext  bool  `json:"has_next"`
}

type Envelope[T any] struct {
	Code    int       `json:"code"`
	Message string    `json:"message"`
	Data    T         `json:"data"`
	Meta    *PageMeta `json:"meta,omitempty"`
}

type ErrorEnvelope struct {
	Code    int     `json:"code"`
	Message string  `json:"message"`
	Data    *string `json:"data"`
}

func OK[T any](data T, meta *PageMeta) Envelope[T] {
	return Envelope[T]{
		Code:    0,
		Message: "success",
		Data:    data,
		Meta:    meta,
	}
}

func WriteOK[T any](w http.ResponseWriter, data T, meta *PageMeta) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(OK(data, meta))
}

func WriteError(w http.ResponseWriter, err *AppError) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(err.StatusCode)
	_ = json.NewEncoder(w).Encode(ErrorEnvelope{
		Code:    err.Code,
		Message: err.Message,
		Data:    nil,
	})
}

